### ISSUE TITLE
Transcription of Write-Information works if preference is 'SilentlyContinue'

### ISSUE THREAD

**Issue by hubuk** [PowerShell/PowerShell#6916]
Related issue: #4645

run.ps1
------------------

```powershell 
Start-Transcript -Path (Join-Path $PSScriptRoot 'transcriptdata.txt')
$InformationPreference = 'SilentlyContinue'
Write-Information 'SilentlyContinue'
$InformationPreference = 'Continue'
Write-Information 'LoudlyContinue'
Stop-Transcript
```

Steps to reproduce
------------------

```powershell 
.\run.ps1
```

Expected behavior
-----------------

```none
transcriptdata.txt should contains only LoudlyContinue
```

Actual behavior
---------------

```none
transcriptdata.txt contains only SilentlyContinue
```

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.1.0-preview.1
PSEdition                      Core
GitCommitId                    v6.1.0-preview.1
OS                             Microsoft Windows 10.0.16299
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

Investigation
----------------

The problem is in if/else statement which starts here:
https://github.com/PowerShell/PowerShell/blob/948532ac839f3bc7e8bbbd3c9f5f292868ac8687/src/System.Management.Automation/engine/MshCommandRuntime.cs#L757-L758
and has a continuation here:
https://github.com/PowerShell/PowerShell/blob/948532ac839f3bc7e8bbbd3c9f5f292868ac8687/src/System.Management.Automation/engine/MshCommandRuntime.cs#L824-L828

Transcription of Write-Host commands are handled in InternalUI.Write* methods as the comment suggests. These commands are distinguished by "PSHOST" tag.
Transcription of Write-Information commands is supposed to take place in the else statement but code in this statement is being executed only if 
``` C#
(!record.Tags.Contains("PSHOST") || record.Tags.Contains("FORWARDED"))
&&
preference != ActionPreference.Continue)
```
This condition is invalid. Correct condition is:
``` C#
!record.Tags.Contains("PSHOST")
&&
preference == ActionPreference.Continue)
```

PR pending...

**Comment by iSazonov** [comment]
@mklement0 Could you please comment?

**Comment by mklement0** [comment]
Thanks for taking this on, @hubuk.

It's definitely a step in the right direction, but other quirks remain, so maybe we can expand the scope of your PR.

The following test code contrast the console output of various `Write-Information` / `Write-Host` / `Out-Host` commands with respect to what prints to the console vs. what shows in the transcript (which should be the same):

```powershell
"--- CONSOLE output`n"
$null = Start-Transcript

Write-Host 'write-host silentlycontinue'
Write-Information 'write-info silentlycontinue'

Write-Host 'write-host continue' -InformationAction Continue
Write-Information 'write-info continue' -InformationAction Continue

Write-Host 'write-host ignore' -InformationAction Ignore
Write-Information 'write-info ignore' -InformationAction Ignore

Out-Host -InputObject 'out-host' # always prints to both the console and the transcript


$f = (Stop-Transcript).Path
"`n---- TRANSCRIPT contents"
((gc -raw $f) -split '(?m)^\*+$')[2]
'----'
Remove-Item $f
```

* Your current PR commendably fixes the `Write-Information` behavior in principle, but in the transcript its output is still prefixed with `INFO: `, unlike in the console.

* `Write-Host` output still _invariably_ end up in the transcript - while `-InformationAction Ignore` / `6>$null` works in the _console_, it doesn't in the transcript.

While fixing the quirks is definitely desirable, we also have to ask ourselves whether user may depend on the old behavior, such as expecting `Write-Information` output to be there by default.

@SteveL-MSFT, any thoughts on this and the other issues mentioned in #4645 with respect to backward compatibility?

**Comment by iSazonov** [comment]
I agree with fixing the Issue too.

For others I'd open new issue for every inconsistency - I believe we get conclusions faster.

**Comment by hubuk** [comment]
@mklement0 I made changes to INFO: prefix, but I observe different behavior in `Write-Host 'write-host ignore' -InformationAction Ignore` than you described. It is present in transcript but not in the console.

```
--- CONSOLE output
write-host silentlycontinue
write-host continue
write-info continue
out-host
---- TRANSCRIPT contents
write-host silentlycontinue
write-host continue
write-info continue
write-host ignore
out-host
----
```

I think that Write-Host should always be visible in both console and transcript regardless of InformationAction and $InformationPreference.

**Comment by mklement0** [comment]
@hubuk:

> but I observe different behavior in Write-Host 'write-host ignore' -InformationAction Ignore than you described. It is present in transcript but not in the console.

Yes, that is the behavior  I was trying to describe.

> I think that Write-Host should always be visible in both console and transcript regardless of InformationAction and $InformationPreference.

That applies to `Out-Host`, but not to `Write-Host`, which is now a bit of a misnomer in that it writes to the information stream, not to the host directly.

From what I understand it is by design that in v5+ you can silence `Write-Host` with ` -InformationAction Ignore` / `6>$null` (and that you can generally redirect it with `6>`).

In short: My understanding is that `Write-Host` is now an output-by-default companion to the silent-by-default `Write-Information` (with additional, UI-related features).

**Comment by iSazonov** [comment]
> Write-Host, which is now a bit of a misnomer in that it writes to the information stream, not to the host directly.

It writes to the informational stream and then to host directly based on tag presence.

**Comment by mklement0** [comment]
@iSazonov: Yes, but you can suppress `Write-Host` output to the _host_ with `-InformationAction Ignore` / `6>$null`, whereas the output still shows in the _transcript_, and that's the problem.

**Comment by iSazonov** [comment]
>the output still shows in the transcript, and that's the problem

Agree.

**Comment by hubuk** [comment]
> the output still shows in the transcript, and that's the problem

If `-InformatioAction Ignore` should suppress `Write-Host` from writing to console then `-InformationAction SilentlyContinue` should also suppress console output for consistency.

But for me it seems more natural that `Write-Host` should always writes to host. Its interaction with InformationStream can be controlled using InformationAction and InformationPreference variable. If the appearance on the console shall be controlled using these then Write-Information should be used instead. That was my understanding of the differences between these two commands.

**Comment by mklement0** [comment]
> If `-InformationAction Ignore` should suppress `Write-Host` from writing to console, then `-InformationAction SilentlyContinue` should also suppress console output for consistency.

That's not possible, because that would make `Write-Host` _silent_ by default, given that `$InformationPreference` defaults to `SilentlyContinue` and, in the absence of an explicit `-InformationAction` argument, is equivalent to passing `-InformationAction SilentlyContinue`.

For backward compatibility (alone) you don't want `Write-Host` to be silent by default.

> But for me it seems more natural that Write-Host should always writes to host.

That's what the name of the cmdlet suggests, and that's how it _used to_ work.

However, as https://github.com/PowerShell/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Utility/Write-Information.md states:

> Starting in Windows PowerShell 5.0, Write-Host is a wrapper for Write-Information. You can now use Write-Host to emit output to the information stream, but the $InformationPreference preference variable and InformationAction common parameter do not affect Write-Host messages. 

(Ignore the fact that this is not the entirely accurate, because the `Ignore` value passed to `-InformationAction` _does_ affect `Write-Host` and is in effect equivalent to `6>$null`.)

Presumably - and perhaps someone from the team can confirm - this slight repurposing of `Write-Host` since v5 was to compensate for frequent misuse of `Write-Host` by cmdlet/function authors, which resulted in callers being unable to capture or suppress its output.

By making `Write-Host` write to the information stream (number `6`) instead, it became possible to either suppress or capture its output. (Note that, as stated, `Out-Host` indeed still only and invariably prints to the host (and possibly transcript)).

**Comment by hubuk** [comment]
Closed via #6917.
